Geyser fails to start w/paper - DnsQuestion class missing from Netty library

### Describe the bug

On 2025/01/17 my instance of Geyser running on paper was working without issue. I am running the server as a docker image (https://github.com/itzg/docker-minecraft-server).

 As of today 2025/01/18 it has stopped working after the Geyser plugin was automatically updated from the latest build [613718a](https://github.com/GeyserMC/Geyser/commit/613718a52d108871efc5569615cc68f93e96098c).

It appears the failure is due to the fact that the following class can no-longer found in the Netty library:
**io.netty.handler.codec.dns.DnsQuestion**

This throws a NoClassDefFoundError in the server startup logs, and also prevents Bedrock players from connecting to the server. See screenshots and additional info for raw error outputs.


### To Reproduce

1. Start up a paper server (1.21.4-118-main@73f7736) running Geyser and Floodgate
2. Wait for the server to load and observe a `java.lang.NoClassDefFoundError: io/netty/handler/codec/dns/DnsQuestion`
3. Try to connect to the server and observe further errors and inability to join.

### Expected behaviour

- No exceptions thrown during server startup. Server was behaving normally on 2025/01/17.
- Ability for players to connect to the server without being thrown an error.

### Screenshots / Videos

![Image](https://github.com/user-attachments/assets/34f6d1ad-31a6-4661-8b23-67715408120b)

### Server Version and Plugins

- Java 21 (OpenJDK 64-Bit Server VM 21.0.5+11-LTS; Eclipse Adoptium Temurin-21.0.5+11) on Linux 6.8.0-1020-gcp (amd64)
- Paper 1.21.4-118-main@73f7736 (2025-01-18T13:37:35Z) for Minecraft 1.21.4
- floodgate v2.2.3-SNAPSHOT (b114-83fbd6e)
- Geyser version 2.6.0-b748 (git-master-613718a)
- no additonal plugins

### Geyser Dump

_No response_

### Geyser Version

2.6.0-b748

### Minecraft: Bedrock Edition Device/Version

1.21.4

### Additional Context
```
[init] Running as uid=1003 gid=1004 with /data as 'drwxrwxrwx 12 1003 1004 4096 Jan 18 13:41 /data'
[init] Resolving type given PAPER
[mc-image-helper] 14:12:35.894 INFO  : The file /data/plugins/floodgate-spigot.jar is already up to date
[mc-image-helper] 14:12:35.894 INFO  : The file /data/plugins/Geyser-Spigot.jar is already up to date
[mc-image-helper] 14:12:37.908 INFO  : Created/updated 1 property in /data/server.properties
[init] Downloading default configs, if needed
[mc-image-helper] 14:12:39.580 INFO  : The file /data/config/paper-world-defaults.yml already exists
[mc-image-helper] 14:12:39.639 INFO  : The file /data/config/paper-global.yml already exists
[init] Downloading default top-level configs, if needed
[mc-image-helper] 14:12:41.682 INFO  : The file /data/bukkit.yml already exists
[mc-image-helper] 14:12:41.679 INFO  : The file /data/spigot.yml already exists
[init] Setting initial memory to 2G and max to 4G
[init] Starting the Minecraft server...
Starting org.bukkit.craftbukkit.Main
2025-01-18T14:12:43.421294092Z ServerMain WARN Advanced terminal features are not available in this environment
[14:12:43 INFO]: [bootstrap] Running Java 21 (OpenJDK 64-Bit Server VM 21.0.5+11-LTS; Eclipse Adoptium Temurin-21.0.5+11) on Linux 6.8.0-1020-gcp (amd64)
[14:12:43 INFO]: [bootstrap] Loading Paper 1.21.4-118-main@73f7736 (2025-01-18T13:37:35Z) for Minecraft 1.21.4
[14:12:43 INFO]: [PluginInitializerManager] Initializing plugins...
[14:12:44 INFO]: [PluginInitializerManager] Initialized 2 plugins
[14:12:44 INFO]: [PluginInitializerManager] Bukkit plugins (2):
 - Geyser-Spigot (2.6.0-SNAPSHOT), floodgate (2.2.3-SNAPSHOT (b114-83fbd6e))
[14:12:49 INFO]: Environment: Environment[sessionHost=https://sessionserver.mojang.com, servicesHost=https://api.minecraftservices.com, name=PROD]
[14:12:51 INFO]: Loaded 1370 recipes
[14:12:51 INFO]: Loaded 1481 advancements
[14:12:51 INFO]: [MCTypeRegistry] Initialising converters for DataConverter...
[14:12:51 INFO]: [MCTypeRegistry] Finished initialising converters for DataConverter in 235.4ms
[14:12:51 INFO]: Starting minecraft server version 1.21.4
[14:12:51 INFO]: Loading properties
[14:12:51 INFO]: This server is running Paper version 1.21.4-118-main@73f7736 (2025-01-18T13:37:35Z) (Implementing API version 1.21.4-R0.1-SNAPSHOT)
[14:12:51 INFO]: [spark] This server bundles the spark profiler. For more information please visit https://docs.papermc.io/paper/profiling
[14:12:51 INFO]: Using 4 threads for Netty based IO
[14:12:51 INFO]: Server Ping Player Sample Count: 12
[14:12:52 INFO]: [MoonriseCommon] Paper is using 1 worker threads, 1 I/O threads
[14:12:52 INFO]: [ChunkTaskScheduler] Chunk system is using population gen parallelism: true
[14:12:53 INFO]: Default game type: SURVIVAL
[14:12:53 INFO]: Generating keypair
[14:12:53 INFO]: Starting Minecraft server on *:25565
[14:12:53 INFO]: Using epoll channel type
[14:12:53 INFO]: Paper: Using libdeflate (Linux x86_64) compression from Velocity.
[14:12:53 INFO]: Paper: Using OpenSSL 3.x.x (Linux x86_64) cipher from Velocity.
[14:12:53 INFO]: [floodgate] Loading server plugin floodgate v2.2.3-SNAPSHOT (b114-83fbd6e)
[14:12:54 INFO]: [floodgate] Took 860ms to boot Floodgate
[14:12:54 INFO]: [Geyser-Spigot] Loading server plugin Geyser-Spigot v2.6.0-SNAPSHOT
[14:12:55 INFO]: [Geyser-Spigot] Loading extensions...
[14:12:55 INFO]: [Geyser-Spigot] Loaded 0 extension(s)
[14:12:55 INFO]: Server permissions file permissions.yml is empty, ignoring it
[14:12:55 INFO]: Preparing level "world"
[14:12:55 INFO]: Preparing start region for dimension minecraft:overworld
[14:12:56 INFO]: Preparing spawn area: 0%
[14:12:56 INFO]: Preparing spawn area: 2%
[14:12:56 INFO]: Preparing spawn area: 20%
[14:12:56 INFO]: Time elapsed: 1157 ms
[14:12:56 INFO]: Preparing start region for dimension minecraft:the_nether
[14:12:57 INFO]: Preparing spawn area: 0%
[14:12:57 INFO]: Time elapsed: 277 ms
[14:12:57 INFO]: Preparing start region for dimension minecraft:the_end
[14:12:57 INFO]: Preparing spawn area: 0%
[14:12:57 INFO]: Time elapsed: 120 ms
[14:12:57 INFO]: [floodgate] Enabling floodgate v2.2.3-SNAPSHOT (b114-83fbd6e)
[14:12:57 INFO]: [Geyser-Spigot] Enabling Geyser-Spigot v2.6.0-SNAPSHOT
[14:12:58 INFO]: [spark] Starting background profiler...
[14:12:58 INFO]: [Geyser-Spigot] ******************************************
[14:12:58 INFO]: [Geyser-Spigot] 
[14:12:58 INFO]: [Geyser-Spigot] Loading Geyser version 2.6.0-b748 (git-master-613718a)
[14:12:58 INFO]: [Geyser-Spigot] 
[14:12:58 INFO]: [Geyser-Spigot] ******************************************
[14:13:04 INFO]: [Geyser-Spigot] Started Geyser on UDP port 19132
[14:13:04 INFO]: [Geyser-Spigot] Done (5.441s)! Run /geyser help for help!
[14:13:04 ERROR]: Could not pass event ServerLoadEvent to Geyser-Spigot v2.6.0-SNAPSHOT
java.lang.NoClassDefFoundError: io/netty/handler/codec/dns/DnsQuestion
        at Geyser-Spigot.jar/org.geysermc.mcprotocollib.network.session.ClientNetworkSession.connect(ClientNetworkSession.java:64) ~[Geyser-Spigot.jar:?]
        at Geyser-Spigot.jar/org.geysermc.mcprotocollib.network.ClientSession.connect(ClientSession.java:13) ~[Geyser-Spigot.jar:?]
        at Geyser-Spigot.jar/org.geysermc.geyser.platform.spigot.GeyserSpigotInjector.workAroundWeirdBug(GeyserSpigotInjector.java:183) ~[Geyser-Spigot.jar:?]
        at Geyser-Spigot.jar/org.geysermc.geyser.platform.spigot.GeyserSpigotInjector.initializeLocalChannel0(GeyserSpigotInjector.java:142) ~[Geyser-Spigot.jar:?]
        at Geyser-Spigot.jar/org.geysermc.geyser.network.netty.GeyserInjector.initializeLocalChannel(GeyserInjector.java:64) ~[Geyser-Spigot.jar:?]
        at Geyser-Spigot.jar/org.geysermc.geyser.platform.spigot.GeyserSpigotPlugin.onGeyserEnable(GeyserSpigotPlugin.java:265) ~[Geyser-Spigot.jar:?]
        at Geyser-Spigot.jar/org.geysermc.geyser.platform.spigot.GeyserSpigotPlugin$1.onServerLoaded(GeyserSpigotPlugin.java:220) ~[Geyser-Spigot.jar:?]
        at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:80) ~[paper-api-1.21.4-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70) ~[paper-api-1.21.4-R0.1-SNAPSHOT.jar:?]
        at io.papermc.paper.plugin.manager.PaperEventManager.callEvent(PaperEventManager.java:54) ~[paper-1.21.4.jar:1.21.4-118-73f7736]
        at io.papermc.paper.plugin.manager.PaperPluginManagerImpl.callEvent(PaperPluginManagerImpl.java:131) ~[paper-1.21.4.jar:1.21.4-118-73f7736]
        at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:628) ~[paper-api-1.21.4-R0.1-SNAPSHOT.jar:?]
        at net.minecraft.server.MinecraftServer.loadWorld0(MinecraftServer.java:750) ~[paper-1.21.4.jar:1.21.4-118-73f7736]
        at net.minecraft.server.MinecraftServer.loadLevel(MinecraftServer.java:488) ~[paper-1.21.4.jar:1.21.4-118-73f7736]
        at net.minecraft.server.dedicated.DedicatedServer.initServer(DedicatedServer.java:322) ~[paper-1.21.4.jar:1.21.4-118-73f7736]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1163) ~[paper-1.21.4.jar:1.21.4-118-73f7736]
        at net.minecraft.server.MinecraftServer.lambda$spin$2(MinecraftServer.java:310) ~[paper-1.21.4.jar:1.21.4-118-73f7736]
        at java.base/java.lang.Thread.run(Unknown Source) ~[?:?]
Caused by: java.lang.ClassNotFoundException: io.netty.handler.codec.dns.DnsQuestion
        at org.bukkit.plugin.java.PluginClassLoader.loadClass0(PluginClassLoader.java:197) ~[paper-api-1.21.4-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.plugin.java.PluginClassLoader.loadClass(PluginClassLoader.java:164) ~[paper-api-1.21.4-R0.1-SNAPSHOT.jar:?]
        at java.base/java.lang.ClassLoader.loadClass(Unknown Source) ~[?:?]
        ... 18 more
[14:13:04 INFO]: Done preparing level "world" (8.927s)
[14:13:04 INFO]: Starting remote control listener
[14:13:04 INFO]: Thread RCON Listener started
[14:13:04 INFO]: RCON running on 0.0.0.0:25575
[14:13:04 INFO]: Running delayed init tasks
[14:13:04 INFO]: Done (21.473s)! For help, type "help"
[14:13:30 WARN]: [Geyser-Spigot] Error occurred while initializing player!
java.lang.NullPointerException: Cannot invoke "org.geysermc.geyser.level.WorldManager.hasOwnChunkCache()" because the return value of "org.geysermc.geyser.GeyserImpl.getWorldManager()" is null
        at Geyser-Spigot.jar/org.geysermc.geyser.session.cache.ChunkCache.<init>(ChunkCache.java:47) ~[Geyser-Spigot.jar:?]
        at Geyser-Spigot.jar/org.geysermc.geyser.session.GeyserSession.<init>(GeyserSession.java:666) ~[Geyser-Spigot.jar:?]
        at Geyser-Spigot.jar/org.geysermc.geyser.network.GeyserServerInitializer.initSession(GeyserServerInitializer.java:65) ~[Geyser-Spigot.jar:?]
        at Geyser-Spigot.jar/org.geysermc.geyser.network.GeyserServerInitializer.initSession(GeyserServerInitializer.java:41) ~[Geyser-Spigot.jar:?]
        at Geyser-Spigot.jar/org.cloudburstmc.protocol.bedrock.netty.initializer.BedrockChannelInitializer.createSession(BedrockChannelInitializer.java:112) ~[Geyser-Spigot.jar:?]
        at Geyser-Spigot.jar/org.geysermc.geyser.network.GeyserServerInitializer.lambda$createPeer$0(GeyserServerInitializer.java:80) ~[Geyser-Spigot.jar:?]
        at Geyser-Spigot.jar/org.cloudburstmc.protocol.bedrock.BedrockPeer.channelActive(BedrockPeer.java:244) ~[Geyser-Spigot.jar:?]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelActive(AbstractChannelHandlerContext.java:260) ~[netty-transport-4.1.115.Final.jar:4.1.115.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelActive(AbstractChannelHandlerContext.java:238) ~[netty-transport-4.1.115.Final.jar:4.1.115.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelActive(AbstractChannelHandlerContext.java:231) ~[netty-transport-4.1.115.Final.jar:4.1.115.Final]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelActive(DefaultChannelPipeline.java:1345) ~[netty-transport-4.1.115.Final.jar:4.1.115.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelActive(AbstractChannelHandlerContext.java:258) ~[netty-transport-4.1.115.Final.jar:4.1.115.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelActive(AbstractChannelHandlerContext.java:238) ~[netty-transport-4.1.115.Final.jar:4.1.115.Final]
        at io.netty.channel.DefaultChannelPipeline.fireChannelActive(DefaultChannelPipeline.java:844) ~[netty-transport-4.1.115.Final.jar:4.1.115.Final]
        at Geyser-Spigot.jar/org.cloudburstmc.netty.handler.codec.raknet.server.RakServerOnlineInitialHandler.lambda$channelRead0$0(RakServerOnlineInitialHandler.java:73) ~[Geyser-Spigot.jar:?]
        at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173) ~[netty-common-4.1.115.Final.jar:4.1.115.Final]
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166) ~[netty-common-4.1.115.Final.jar:4.1.115.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472) ~[netty-common-4.1.115.Final.jar:4.1.115.Final]
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:405) ~[netty-transport-classes-epoll-4.1.115.Final.jar:4.1.115.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[netty-common-4.1.115.Final.jar:4.1.115.Final]
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.115.Final.jar:4.1.115.Final]
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.115.Final.jar:4.1.115.Final]
        at java.base/java.lang.Thread.run(Unknown Source) ~[?:?]
```

--- a/protocol/src/main/java/org/geysermc/mcprotocollib/network/helper/NettyHelper.java
+++ b/protocol/src/main/java/org/geysermc/mcprotocollib/network/helper/NettyHelper.java
@@ -12,6 +12,8 @@ import io.netty.handler.codec.haproxy.HAProxyProxiedProtocol;
 import io.netty.handler.proxy.HttpProxyHandler;
 import io.netty.handler.proxy.Socks4ProxyHandler;
 import io.netty.handler.proxy.Socks5ProxyHandler;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
 import org.geysermc.mcprotocollib.network.BuiltinFlags;
 import org.geysermc.mcprotocollib.network.ProxyInfo;
 import org.geysermc.mcprotocollib.network.Session;
@@ -29,7 +31,6 @@ import java.net.InetSocketAddress;
 import java.net.SocketAddress;
 import java.net.UnknownHostException;
 import java.util.Hashtable;
-import java.util.Optional;
 
 public class NettyHelper {
     private static final Logger log = LoggerFactory.getLogger(NettyHelper.class);
@@ -50,22 +51,24 @@ public class NettyHelper {
         }
     }
 
-    public static SocketAddress resolveAddress(Session session, SocketAddress address) {
+    public static @NonNull SocketAddress resolveAddress(Session session, SocketAddress address) {
         if (address instanceof InetSocketAddress inetAddress && inetAddress.isUnresolved()) {
-            return resolveAddress(session, inetAddress.getHostString(), inetAddress.getPort())
-                .orElse(address);
+            SocketAddress resolved = resolveAddress(session, inetAddress.getHostString(), inetAddress.getPort());
+            if (resolved != null) {
+                return resolved;
+            }
         }
 
         return address;
     }
 
-    public static Optional<SocketAddress> resolveAddress(Session session, String host, int port) {
+    public static @Nullable SocketAddress resolveAddress(Session session, String host, int port) {
         ServerAddress serverAddress = ServerAddress.fromStringAndPort(host, port);
 
         if (session.getFlag(BuiltinFlags.ATTEMPT_SRV_RESOLVE, true) && serverAddress.port() == MC_JAVA_DEFAULT_PORT) {
             // SRVs can override address on Java, but not Bedrock.
-            Optional<SocketAddress> resolved = resolveSrv(session, serverAddress);
-            if (resolved.isPresent()) {
+            SocketAddress resolved = resolveSrv(session, serverAddress);
+            if (resolved != null) {
                 return resolved;
             }
         } else {
@@ -75,7 +78,7 @@ public class NettyHelper {
         return resolveByHost(serverAddress);
     }
 
-    private static Optional<SocketAddress> resolveSrv(Session session, ServerAddress serverAddress) {
+    private static @Nullable SocketAddress resolveSrv(Session session, ServerAddress serverAddress) {
         String name = session.getPacketProtocol().getSRVRecordPrefix() + "._tcp." + serverAddress.host();
         log.debug("Attempting SRV lookup for \"{}\".", name);
 
@@ -94,18 +97,18 @@ public class NettyHelper {
             log.debug("Failed to resolve SRV record.", e);
         }
 
-        return Optional.empty();
+        return null;
     }
 
-    private static Optional<SocketAddress> resolveByHost(ServerAddress serverAddress) {
+    private static @Nullable SocketAddress resolveByHost(ServerAddress serverAddress) {
         try {
             String host = serverAddress.host();
             InetAddress resolved = InetAddress.getByName(host);
             log.debug("Resolved {} -> {}", host, resolved.getHostAddress());
-            return Optional.of(new InetSocketAddress(resolved, serverAddress.port()));
+            return new InetSocketAddress(resolved, serverAddress.port());
         } catch (UnknownHostException e) {
             log.debug("Failed to resolve host.", e);
-            return Optional.empty();
+            return null;
         }
     }
 

--- a/protocol/src/main/java/org/geysermc/mcprotocollib/network/helper/NettyHelper.java
+++ b/protocol/src/main/java/org/geysermc/mcprotocollib/network/helper/NettyHelper.java
@@ -40,7 +40,7 @@ public class NettyHelper {
         try {
             String contextFactory = "com.sun.jndi.dns.DnsContextFactory";
             Class.forName(contextFactory);
-            Hashtable<String, String> environment = new Hashtable<String, String>();
+            Hashtable<String, String> environment = new Hashtable<>();
             environment.put("java.naming.factory.initial", contextFactory);
             environment.put("java.naming.provider.url", "dns:");
             environment.put("com.sun.jndi.dns.timeout.retries", "1");

--- a/protocol/src/main/java/org/geysermc/mcprotocollib/network/helper/NettyHelper.java
+++ b/protocol/src/main/java/org/geysermc/mcprotocollib/network/helper/NettyHelper.java
@@ -1,18 +1,9 @@
 package org.geysermc.mcprotocollib.network.helper;
 
-import io.netty.buffer.ByteBuf;
-import io.netty.channel.AddressedEnvelope;
 import io.netty.channel.Channel;
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.channel.ChannelInboundHandlerAdapter;
 import io.netty.channel.ChannelPipeline;
-import io.netty.channel.EventLoop;
-import io.netty.handler.codec.dns.DefaultDnsQuestion;
-import io.netty.handler.codec.dns.DefaultDnsRawRecord;
-import io.netty.handler.codec.dns.DefaultDnsRecordDecoder;
-import io.netty.handler.codec.dns.DnsRecordType;
-import io.netty.handler.codec.dns.DnsResponse;
-import io.netty.handler.codec.dns.DnsSection;
 import io.netty.handler.codec.haproxy.HAProxyCommand;
 import io.netty.handler.codec.haproxy.HAProxyMessage;
 import io.netty.handler.codec.haproxy.HAProxyMessageEncoder;
@@ -21,88 +12,110 @@ import io.netty.handler.codec.haproxy.HAProxyProxiedProtocol;
 import io.netty.handler.proxy.HttpProxyHandler;
 import io.netty.handler.proxy.Socks4ProxyHandler;
 import io.netty.handler.proxy.Socks5ProxyHandler;
-import io.netty.resolver.dns.DnsNameResolver;
-import io.netty.resolver.dns.DnsNameResolverBuilder;
 import org.geysermc.mcprotocollib.network.BuiltinFlags;
 import org.geysermc.mcprotocollib.network.ProxyInfo;
 import org.geysermc.mcprotocollib.network.Session;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import javax.naming.directory.DirContext;
+import javax.naming.directory.InitialDirContext;
+import java.net.IDN;
 import java.net.Inet4Address;
 import java.net.Inet6Address;
 import java.net.InetAddress;
 import java.net.InetSocketAddress;
 import java.net.SocketAddress;
 import java.net.UnknownHostException;
-import java.util.function.Supplier;
+import java.util.Hashtable;
+import java.util.Optional;
 
 public class NettyHelper {
     private static final Logger log = LoggerFactory.getLogger(NettyHelper.class);
-    private static final String IP_REGEX = "\\b\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\b";
+    public static final int MC_JAVA_DEFAULT_PORT = 25565;
+    private static final DirContext DIR_CONTEXT;
 
-    public static SocketAddress resolveAddress(Session session, Supplier<EventLoop> eventLoop, SocketAddress address) {
+    static {
+        try {
+            var contextFactory = "com.sun.jndi.dns.DnsContextFactory";
+            Class.forName(contextFactory);
+            var environment = new Hashtable<String, String>();
+            environment.put("java.naming.factory.initial", contextFactory);
+            environment.put("java.naming.provider.url", "dns:");
+            environment.put("com.sun.jndi.dns.timeout.retries", "1");
+            DIR_CONTEXT = new InitialDirContext(environment);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static SocketAddress resolveAddress(Session session, SocketAddress address) {
         if (address instanceof InetSocketAddress inetAddress && inetAddress.isUnresolved()) {
-            return resolveAddress(session, eventLoop.get(), inetAddress.getHostString(), inetAddress.getPort());
+            return resolveAddress(session, inetAddress.getHostString(), inetAddress.getPort())
+                .orElse(address);
         }
 
         return address;
     }
 
-    public static SocketAddress resolveAddress(Session session, EventLoop eventLoop, String host, int port) {
-        String name = session.getPacketProtocol().getSRVRecordPrefix() + "._tcp." + host;
-        log.debug("Attempting SRV lookup for \"{}\".", name);
+    public static Optional<SocketAddress> resolveAddress(Session session, String host, int port) {
+        ServerAddress serverAddress = ServerAddress.fromStringAndPort(host, port);
 
-        if (session.getFlag(BuiltinFlags.ATTEMPT_SRV_RESOLVE, true) && (!host.matches(IP_REGEX) && !host.equalsIgnoreCase("localhost"))) {
-            try (DnsNameResolver resolver = new DnsNameResolverBuilder(eventLoop)
-                .channelFactory(TransportHelper.TRANSPORT_TYPE.datagramChannelFactory())
-                .build()) {
-                AddressedEnvelope<DnsResponse, InetSocketAddress> envelope = resolver.query(new DefaultDnsQuestion(name, DnsRecordType.SRV)).get();
-                try {
-                    DnsResponse response = envelope.content();
-                    if (response.count(DnsSection.ANSWER) > 0) {
-                        DefaultDnsRawRecord record = response.recordAt(DnsSection.ANSWER, 0);
-                        if (record.type() == DnsRecordType.SRV) {
-                            ByteBuf buf = record.content();
-                            buf.skipBytes(4); // Skip priority and weight.
-
-                            int tempPort = buf.readUnsignedShort();
-                            String tempHost = DefaultDnsRecordDecoder.decodeName(buf);
-                            if (tempHost.endsWith(".")) {
-                                tempHost = tempHost.substring(0, tempHost.length() - 1);
-                            }
-
-                            log.debug("Found SRV record containing \"{}:{}\".", tempHost, tempPort);
-
-                            host = tempHost;
-                            port = tempPort;
-                        } else {
-                            log.debug("Received non-SRV record in response.");
-                        }
-                    } else {
-                        log.debug("No SRV record found.");
-                    }
-                } finally {
-                    envelope.release();
-                }
-            } catch (Exception e) {
-                log.debug("Failed to resolve SRV record.", e);
+        if (session.getFlag(BuiltinFlags.ATTEMPT_SRV_RESOLVE, true) && serverAddress.port() == MC_JAVA_DEFAULT_PORT) {
+            // SRVs can override address on Java, but not Bedrock.
+            var resolved = resolveSrv(session, serverAddress);
+            if (resolved.isPresent()) {
+                return resolved;
             }
         } else {
-            log.debug("Not resolving SRV record for {}", host);
+            log.debug("Not resolving SRV record for {}", serverAddress.host());
         }
 
-        // Resolve host here
+        return resolveByHost(serverAddress);
+    }
+
+    private static Optional<SocketAddress> resolveSrv(Session session, ServerAddress serverAddress) {
+        String name = session.getPacketProtocol().getSRVRecordPrefix() + "._tcp." + serverAddress.host();
+        log.debug("Attempting SRV lookup for \"{}\".", name);
+
         try {
-            InetAddress resolved = InetAddress.getByName(host);
+            var srvAttribute = DIR_CONTEXT.getAttributes(name, new String[]{"SRV"}).get("srv");
+            if (srvAttribute != null) {
+                var attributeSplit = srvAttribute.get().toString().split(" ", 4);
+                log.debug("SRV lookup resolved \"{}\" to \"{}\".", name, srvAttribute.get().toString());
+
+                return resolveByHost(
+                    ServerAddress.fromStringAndPort(attributeSplit[3], parseJavaPort(attributeSplit[2])));
+            } else {
+                log.debug("SRV lookup for \"{}\" returned no records.", name);
+            }
+        } catch (Exception e) {
+            log.debug("Failed to resolve SRV record.", e);
+        }
+
+        return Optional.empty();
+    }
+
+    private static Optional<SocketAddress> resolveByHost(ServerAddress serverAddress) {
+        try {
+            var host = serverAddress.host();
+            var resolved = InetAddress.getByName(host);
             log.debug("Resolved {} -> {}", host, resolved.getHostAddress());
-            return new InetSocketAddress(resolved, port);
+            return Optional.of(new InetSocketAddress(resolved, serverAddress.port()));
         } catch (UnknownHostException e) {
-            log.debug("Failed to resolve host, letting Netty do it instead.", e);
-            return InetSocketAddress.createUnresolved(host, port);
+            log.debug("Failed to resolve host.", e);
+            return Optional.empty();
         }
     }
-    
+
+    private static int parseJavaPort(String port) {
+        try {
+            return Integer.parseInt(port);
+        } catch (NumberFormatException e) {
+            return MC_JAVA_DEFAULT_PORT;
+        }
+    }
+
     public static void initializeHAProxySupport(Session session, Channel channel) {
         InetSocketAddress clientAddress = session.getFlag(BuiltinFlags.CLIENT_PROXIED_ADDRESS);
         if (clientAddress == null) {
@@ -188,6 +201,24 @@ public class NettyHelper {
                 }
             }
             default -> throw new UnsupportedOperationException("Unsupported proxy type: " + proxy.type());
+        }
+    }
+
+    private record ServerAddress(String internalHost, int internalPort) {
+        public static ServerAddress fromStringAndPort(String host, int port) {
+            return new ServerAddress(host, port);
+        }
+
+        public String host() {
+            try {
+                return IDN.toASCII(internalHost);
+            } catch (IllegalArgumentException e) {
+                return "";
+            }
+        }
+
+        public int port() {
+            return internalPort;
         }
     }
 }

--- a/protocol/src/main/java/org/geysermc/mcprotocollib/network/helper/NettyHelper.java
+++ b/protocol/src/main/java/org/geysermc/mcprotocollib/network/helper/NettyHelper.java
@@ -39,7 +39,9 @@ public class NettyHelper {
 
     static {
         try {
+            // This code is taken from Minecraft Java Editions ServerRedirectHandler
             String contextFactory = "com.sun.jndi.dns.DnsContextFactory";
+            // Ensure the DNS context factory is available
             Class.forName(contextFactory);
             Hashtable<String, String> environment = new Hashtable<>();
             environment.put("java.naming.factory.initial", contextFactory);

--- a/protocol/src/main/java/org/geysermc/mcprotocollib/network/helper/NettyHelper.java
+++ b/protocol/src/main/java/org/geysermc/mcprotocollib/network/helper/NettyHelper.java
@@ -18,6 +18,7 @@ import org.geysermc.mcprotocollib.network.Session;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import javax.naming.directory.Attribute;
 import javax.naming.directory.DirContext;
 import javax.naming.directory.InitialDirContext;
 import java.net.IDN;
@@ -37,9 +38,9 @@ public class NettyHelper {
 
     static {
         try {
-            var contextFactory = "com.sun.jndi.dns.DnsContextFactory";
+            String contextFactory = "com.sun.jndi.dns.DnsContextFactory";
             Class.forName(contextFactory);
-            var environment = new Hashtable<String, String>();
+            Hashtable<String, String> environment = new Hashtable<String, String>();
             environment.put("java.naming.factory.initial", contextFactory);
             environment.put("java.naming.provider.url", "dns:");
             environment.put("com.sun.jndi.dns.timeout.retries", "1");
@@ -63,7 +64,7 @@ public class NettyHelper {
 
         if (session.getFlag(BuiltinFlags.ATTEMPT_SRV_RESOLVE, true) && serverAddress.port() == MC_JAVA_DEFAULT_PORT) {
             // SRVs can override address on Java, but not Bedrock.
-            var resolved = resolveSrv(session, serverAddress);
+            Optional<SocketAddress> resolved = resolveSrv(session, serverAddress);
             if (resolved.isPresent()) {
                 return resolved;
             }
@@ -79,9 +80,9 @@ public class NettyHelper {
         log.debug("Attempting SRV lookup for \"{}\".", name);
 
         try {
-            var srvAttribute = DIR_CONTEXT.getAttributes(name, new String[]{"SRV"}).get("srv");
+            Attribute srvAttribute = DIR_CONTEXT.getAttributes(name, new String[]{"SRV"}).get("srv");
             if (srvAttribute != null) {
-                var attributeSplit = srvAttribute.get().toString().split(" ", 4);
+                String[] attributeSplit = srvAttribute.get().toString().split(" ", 4);
                 log.debug("SRV lookup resolved \"{}\" to \"{}\".", name, srvAttribute.get().toString());
 
                 return resolveByHost(
@@ -98,8 +99,8 @@ public class NettyHelper {
 
     private static Optional<SocketAddress> resolveByHost(ServerAddress serverAddress) {
         try {
-            var host = serverAddress.host();
-            var resolved = InetAddress.getByName(host);
+            String host = serverAddress.host();
+            InetAddress resolved = InetAddress.getByName(host);
             log.debug("Resolved {} -> {}", host, resolved.getHostAddress());
             return Optional.of(new InetSocketAddress(resolved, serverAddress.port()));
         } catch (UnknownHostException e) {

--- a/protocol/src/main/java/org/geysermc/mcprotocollib/network/session/ClientNetworkSession.java
+++ b/protocol/src/main/java/org/geysermc/mcprotocollib/network/session/ClientNetworkSession.java
@@ -61,7 +61,7 @@ public class ClientNetworkSession extends NetworkSession implements ClientSessio
             .channelFactory(getChannelFactory())
             .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, getFlag(BuiltinFlags.CLIENT_CONNECT_TIMEOUT, 30) * 1000)
             .group(eventLoopGroup)
-            .remoteAddress(NettyHelper.resolveAddress(this, eventLoopGroup::next, remoteAddress))
+            .remoteAddress(NettyHelper.resolveAddress(this, remoteAddress))
             .localAddress(bindAddress)
             .handler(getChannelHandler());
 

